File: 5.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-

class ListNode:
    def __init__(self, data=None, next=None):
        self.data = data
        self.next = next

def reverse1(node):
    """
        用栈的方法
    """
    if node == None or node.next == None:
        return node
    # 用python的列表表示栈
    stack = []
    cur = node
    # 入栈
    while cur:
        stack.append(cur.data)
        cur = cur.next
    # 出栈
    while stack:
        print(stack.pop())

def reverse2(node):
    """
        递归的方式
    """
    if node == None:
        return None
    # 递归基,就算不需要返回值,递归基这里也需要终止函数!
    if node.next == None:
        print(node.data)
        return;
    reverse2(node.next)
    print(node.data)


if __name__ == "__main__":
    dummy = ListNode('begin')
    node = dummy
    for i in range(10):
        node.next = ListNode(data=i)
        node = node.next
    # reverse1(dummy.next)    
    reverse2(dummy.next)

File: 5.cpp
/*
**  5.从尾到头打印链表
**  输入一个链表的头结点,从尾到头反过来打印出每个节点的值
*/

#include <iostream>
#include <stack>
using namespace std;

struct ListNode {
    int       value;
    ListNode* next;
};

void PrintListReversingly_Iteratively(ListNode* pHead) {
    stack<ListNode*> nodes;
    ListNode* pNode = pHead;
    while (pNode->next != NULL) {
        nodes.push(pNode);
        pNode = pNode->next;
    }
    while (!nodes.empty()) {
        pNode = nodes.top();
        nodes.pop();
        cout << pNode->value << " ";
    }
    cout << endl;
}

void PrintListReversingly_Recursively(ListNode* pHead) {
    if (pHead == NULL) return;
    if (pHead->next == NULL) {
        cout << pHead->value << endl;
        return;
    }
    PrintListReversingly_Recursively(pHead->next);
    cout << pHead->value << endl;
}

int main()
{   
    // 初始化一个测试链表
    ListNode* dummy = new ListNode();
    dummy->value = 666; // 开始
    ListNode* pNode = dummy;
    for (int i=0; i<10; i++) {
        ListNode* tmp = new ListNode();
        tmp->value = i;
        pNode->next = tmp;
        pNode = pNode->next;
    }
    // PrintListReversingly_Iteratively(dummy->next);
    PrintListReversingly_Recursively(dummy->next);
    return 0;
}
